### PR TITLE
feat(TDKN-288): Add new "custom audit backend" feature to daikon-audit

### PR DESCRIPTION
**Pull request by anicolas**
**What is the problem this Pull Request is trying to solve?**

This PR allows daikon-audit users to dynamically instantiate a custom audit backend using configuration properties (it was previously only possible through the API).

This enables more flexibility, since it allows to switch backend dynamically (for example in DI jobs for Observability enterprise project).

**What is the chosen solution to this problem?**
 
This feature introduces a new type of backend to daikon-audit: "custom", and a new "backend class name" property, so that the audit backend can be dynamically specified using properties.

**Why introduce this feature in an 1.x branch?**

Because we need this feature for the Observability enterprise project, which has some impacts on DI jobs. Those jobs are currently using the `1.8.0` version of daikon-audit and we don't want to impose a migration (with breaking changes) to daikon 2.x. It should be done at a later time.

But since this feature is useful, it will be backported to daikon 2.x and 3.x branches:

* Report to 2.x branch: https://jira.talendforge.org/browse/TDKN-289
* Report to 3.x branch: https://jira.talendforge.org/browse/TDKN-290

**Link to the JIRA issue**

* Main JIRA: https://jira.talendforge.org/browse/TDKN-287
* 1.x implementation: https://jira.talendforge.org/browse/TDKN-288
 
**Please check if the Pull Request fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/Talend/daikon/blob/master/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features, coverage should be over 75% in the new code)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in JIRA), if any, are all linked or available in the Pull Request

 
**[ ] This Pull Request introduces a breaking change**

### CODE REVIEW

**Comment by anicolas** [comment]
@fhuaulme the ability to change audit backend doesn't seem to be documented. Should I add some documentation or is it something we'll do at a later time (never? 😄 ).

**Comment by anicolas** [comment]
There seems to be test failures on `daikon-tql-mongo` 🤔 

It seems it can't connect to MongoDB... I'll try to investigate but I'll probably need some help on this one.

File: daikon-audit/audit-common/src/main/java/org/talend/logging/audit/impl/DefaultAuditLoggerBase.java [hunk 1]
```diff
@@ -60,6 +60,19 @@ public DefaultAuditLoggerBase(AuditConfigurationMap externalConfig) {
             this.logger = loadBackend(LOG4J2_BACKEND, config);
             break;
 
+        case CUSTOM:
+            final String backendClass = AuditConfiguration.BACKEND_CLASS_NAME.getString(config);
+
+            if (backendClass == null) {
+                throw new IllegalArgumentException("Selected backend is " + backend + " and no backend class name has been specified");
```
Comment (fhuaulme) [inline daikon-audit/audit-common/src/main/java/org/talend/logging/audit/impl/DefaultAuditLoggerBase.java:67]: Please change message to `"Selected backend is '" + backend + "' and...` (add the surrounding single quote: this eases to detect empty/whitespace issues in error messages)

File: daikon-audit/audit-common/src/test/java/org/talend/logging/audit/custom/CustomAuditBackend.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.slf4j.MDC;
+import org.talend.logging.audit.LogLevel;
+import org.talend.logging.audit.impl.AbstractBackend;
+import org.talend.logging.audit.impl.AuditConfigurationMap;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class CustomAuditBackend extends AbstractBackend {
```
Comment (fhuaulme) [inline daikon-audit/audit-common/src/test/java/org/talend/logging/audit/custom/CustomAuditBackend.java:12]: A quick javadoc would be nice here (with mention of the load of the custom class by its name).

File: daikon-audit/audit-common/src/main/java/org/talend/logging/audit/impl/DefaultAuditLoggerBase.java (hunk 1 above)
Comment (anicolas) [reply daikon-audit/audit-common/src/main/java/org/talend/logging/audit/impl/DefaultAuditLoggerBase.java:67]: It makes sense, but I don't know if you noticed: I simply kept the existing wording used for other error cases in this method.

Would you like me to make the change anyway?

Comment (fhuaulme) [reply daikon-audit/audit-common/src/main/java/org/talend/logging/audit/impl/DefaultAuditLoggerBase.java:67]: I honestly didn't check the previous message (only saw your change). Yes, please do the change as it improves a bit the existing code.

File: daikon-audit/audit-common/src/test/java/org/talend/logging/audit/custom/CustomAuditBackend.java (hunk 2 above)
Comment (anicolas) [reply daikon-audit/audit-common/src/test/java/org/talend/logging/audit/custom/CustomAuditBackend.java:27]: Good idea 👍

**Pull request merged**